pkg/operator/controller/controller_dns_configmap.go

```diff
@@ -270,6 +234,50 @@ func desiredDNSConfigMap(dns *operatorv1.DNS, clusterDomain string, caBundleRevi
 	return cm, nil
 }
 
+// validateTLSSettings validates if serverName is set when tls is enabled, and validates if tls is configured only for ip addresses.
+// If a serverName is set, but tls is not enabled, then the serverName is ignored.
+func validateTLSSettings(dns *operatorv1.DNS) (*operatorv1.DNS, error) {
+	updated := dns.DeepCopy()
+	for i, server := range updated.Spec.Servers {
+		transport := server.ForwardPlugin.TransportConfig.Transport
+		tls := server.ForwardPlugin.TransportConfig.TLS
+		if transport == operatorv1.TLSTransport {
+			// tls cannot be configured without a ServerName
+			if tls == nil || tls.ServerName == "" {
+				return updated, errTransportTLSConfiguredWithoutServerName
+			}
+			// tls can only be configured for ip addresses
+			for _, upstream := range server.ForwardPlugin.Upstreams {
+				if _, _, err := net.SplitHostPort(upstream); err != nil {
+					if res := net.ParseIP(upstream); res == nil {
+						return updated, errTransportTLSConfiguredForNonIP
```

Comment:
to be clear, I think this is a good error message if I understand the context correctly.